### DISCUSSION TITLE
Get a resource specific hook from Connection.get_hook() for AWS connections

### DISCUSSION THREAD

**Discussion by cansjt** [apache/airflow#22820] [Q&A] [answered]
We'd like to be able to create a specific resource specific hook from a AWS Connection. As of yet, the `Connection.get_hook()` method, does not take any parameters and hence the methods fails with the following error:
```
Traceback (most recent call last):
  File "~/project/script.py", line 137, in <module>
    engine = configuration.connection_retriever.get_engine(ATHENA_CONN_ID)
  File "~/project/integrations/airflow/connection.py", line 37, in get_engine
    hook = connection.get_hook()
  File "~/.virtualenvs/project/lib/python3.9/site-packages/airflow/models/connection.py", line 311, in get_hook
    return hook_class(**{conn_id_param: self.conn_id})
  File "~/.virtualenvs/project/lib/python3.9/site-packages/airflow/providers/amazon/aws/hooks/base_aws.py", line 399, in __init__
    if not (self.client_type or self.resource_type):
airflow.exceptions.AirflowException: Either client_type or resource_type must be provided.
```
I cannot see a way to pass those informations to the `get_hook()` method. E.g. add the `client_type` needed to the connection _extras_. But without the means to pass that information to the method, that's pointless. Am I missing something.

Wierdly on master changes have been made to allow passing arguments to `Connection.get_hook()`. See for example [this merged PR](https://github.com/apache/airflow/pull/19849/files). But apparently never released. Is this normal?

It sure makes the `Connection.get_hook()` API less than useful, when working with AWS. I was looking in the code, and especially in the `ProvideManager()` for possible other API to achieve a similar goal. But found nothing yet. Any directions someone could give, to solve this?

Thanks!

**Comment by potiuk** [comment, marked as answer]
Yes. This is normal. get_hook with Kwargs is going to be released as new feature in Airlfow 2.3.0 - this is a new feature that's why it has not been released in 2.2.* bugfix releases, There is nothing weird that stuff added in `main` branch is released when there is new release. 

What else did you expect? What would be non-weird if I may ask?

And yes. it's been done to handle similar cases as this one. For now you can just instantiate the Hook by directly creating it like  in Athena Operators.

**Reply by cansjt** [reply]
It's just that a long time has passed since the [PR has been merged](https://github.com/apache/airflow/pull/18718). And because this change seems to be quite an enabler for different providers (AWS, Snowflake, BigQuery and, I would expect, many providers that have a hierarchy of hook classes and which base class is "abstract").

And sadly, the solution you suggest does not provide the level of abstraction I need/wish for. I cannot/do not want to have to import a provider specific hook class, in the module where this code is used.

But I guess that is the best answer I can hope for, until 2.3.0. Thanks.

**Reply by potiuk** [reply]
We strictly follow SemVer. By deffinition we cannot relase new features in 2.2.* line. Any new features will come in 2.3. This is a new feature. The unluckly thing was that it was indeed merged few days after 2.2.0 has been released  - which made it wait so long.

We are however talking about speeding up new feature releases - basically faster cadence of those - so maybe in the future it will be faster to get new things.

On the other hands MANY customers complains that Airflow speed of development is far too fast for them, They are not prepared to make heavy migration every month. So this is really a balance of stability vs new features that needs to be taken into account.

Sometimes you just need to wait. You can also - if you want - port some features from the "to-be-releaed" Airflow. Airflow is open-source. You are free to modify your code and bring in upstream changes - especially that you know they will be released. So if you need somethig badly, there is nothing preventing you to cherry-pick the changes you want.

Hell - you are even encouraged to modify your Airflow and bring back and contribute your changes back.

Yes - you will not get community support and strength for that (until we release a new feature version), but this is the price to pay.

It's open-source. It's your choice. You  have great powers to exercise, but when you do - it comes with responsibility.

**Reply by aspain** [reply]
@cansjt can I ask what value you provided in the extras field? I'm am getting the same error setting up an aws connection 

> E.g. add the client_type needed to the connection extras

**Reply by cansjt** [reply]
I didn't provide anything. It was a suggestion/idea, but the fact is: the API does not support this yet, so I gave up using Airflow's API. Which given the project I am working on is not that much of an issue. I cannot tell if that is a viable option for you.
